tiflash resource control design doc (#8125)
pingcap/tiflash: docs/design/2023-09-21-tiflash-resource-control.md

# TiFlash Resource Control
## Introduction
This document introduces the design and implementation of TiFlash's resource management feature.

## Motivation and Background
TiDB currently supports global resource management to support multiple businesses in a cluster and achieve isolation between them.

As part of the TiDB architecture, TiFlash also needs to support resource management to achieve the capability of isolating different businesses, especially in AP/HTAP scenarios.

## Detailed Design
The basic idea is similar to TiDB/TiKV's resource control. It adopts a combination of rate limiting and priority scheduling:
1. Rate Limiting: When scheduling tasks, it limits them based on the Token Bucket algorithm.
2. Priority: When physical resources are scarce, tasks are scheduled based on their priority, determining which task should run first.

> The underlying idea behind this design is as follows: If the `RU_PER_SEC` setting for each resource group is relatively small, and the sum of these settings does not exceed the real physical resource limit, rate limiting plays the primary role in throttling each resource group. However, to ensure that resources are used more efficiently and to prevent situations where a small `RU_PER_SEC` setting leads to unused physical resources, TiDB's resource management mechanism allows oversubscription, meaning that the total `RU_PER_SEC` values can exceed the physical resources. In such cases, priority scheduling comes into play to prevent resource contention among resource groups.

Reviewer: What does "relatively small" mean here?

Author: 1 RU means 3ms cpu or 64KiB read bytes. If `RU_PER_SEC` is `relatively small`, it means the physical resources are  sufficient for all currenty queries to use.

Reviewer: Does it mean "the sum of these settings does not exceed the real physical resource limit"? If so, I'd suggest remove the "relatively small" statement.

Author: how about:
```
If the `RU_PER_SEC` setting for each resource group is relatively small(a.k.a. the sum of these settings does not exceed the real physical resource limit), then rate limiting plays the primary role in throttling each resource group.
```

Author: done

When TaskScheduler retrieves a task to execute from the ResourceControlQueue, it gives priority to resource groups with higher priorities. Additionally, to achieve flow control, if the tokens for the corresponding resource group have been depleted, it will wait until tokens are refilled.

ResourceControlQueue uses a MultiLevelFeedbackQueue to store all pipeline tasks for a specific resource group. After obtaining the MultiLevelFeedbackQueue corresponding to a specific resource group based on priority strategy, calling its take() method can fetch the specific task.

##### Updating Resource Usage
After TaskScheduler obtains a task, it will execute for 100ms and update its CPU usage status. During this time, it will call relevant interfaces of the LocalAdmissionController (LAC) to update token consumption.

### LocalAdmissionController (LAC)
LAC is responsible for managing the metadata of all resource groups on a TiFlash node, including:
1. Recording the priorities, configurations, and TokenBuckets of all currently known resource groups.

> 2. Watching GAC etcd, updating resource group configurations and detecting resource group deletions.

Reviewer: why does it detect deletions only? Does creation be support automatically?

Author: When the first query of specific resource comes to TiFlash, TiFlash will create it. Etcd only take cares of deletion and config update of resource group .

Reviewer: Maybe we can describe the process of "new resource group creation" somewhere. Just a personal suggestion.

Author: done

### LAC Test
1. Create two resource groups (rg1, rg2), both with relatively large `RU_PER_SEC`, e.g., 3000, run a tpch workload, observe CPU usage, and query execution times. Expect that the query execution times of the two resource groups are roughly the same.
2. Change the `RU_PER_SEC` of rg1 to 1000, observe CPU usage, and query execution times. Expect reduced CPU usage and increased query execution times for rg1.
3. Change the `RU_PER_SEC` of rg2 to 1000, observe CPU usage, and query execution times. Expect reduced CPU usage and increased query execution times for rg2.
4. Set the burstable attribute of rg1 to true, observe CPU usage, and query execution times. Expect increased CPU usage and significantly reduced query execution times for rg1.
5. Test that LAC reports accurate RU usage to GAC.
6. Test that the conversion between the three TokenBucket modes works as expected.
7. Test with multiple TiFlash nodes, running the above tests (1/2/3) with tpch workload, expecting similar results.
8. Test repeated deletion and reconstruction of resource groups. Expect queries to fail with "resource group not found" errors.

## Impacts & Risks

> 1. The introduction of ResourceControlQueue may lead to performance degradation as TaskScheduler's scheduling adds an extra layer of TaskQueue.

Reviewer: In which situation, resource control will be enabled in TiFlash. Do we have any switch to control it?

Author: Yes, I will add it

Author: done

When TaskScheduler retrieves a task to execute from the ResourceControlQueue, it gives priority to resource groups with higher priorities. Additionally, to achieve flow control, if the tokens for the corresponding resource group have been depleted, it will wait until tokens are refilled.

ResourceControlQueue uses a MultiLevelFeedbackQueue to store all pipeline tasks for a specific resource group. After obtaining the MultiLevelFeedbackQueue corresponding to a specific resource group based on priority strategy, calling its take() method can fetch the specific task.

##### Updating Resource Usage
After TaskScheduler obtains a task, it will execute for 100ms and update its CPU usage status. During this time, it will call relevant interfaces of the LocalAdmissionController (LAC) to update token consumption.

### LocalAdmissionController (LAC)
LAC is responsible for managing the metadata of all resource groups on a TiFlash node, including:
1. Recording the priorities, configurations, and TokenBuckets of all currently known resource groups.

> 2. Watching GAC etcd, updating resource group configurations and detecting resource group deletions.(Creation is not managed by watching etcd, it will be created when the first query comes to TiFlash)

Author: ```suggestion
2. Watching GAC etcd, updating resource group configurations and detecting resource group deletions.(The creation of the resource group does not rely on the watching etcd mechanism. Instead, it is automatically created after the first query for that resource group is dispatched to TiFlash.)
```

> 2. Watching GAC etcd, updating resource group configurations and detecting resource group deletions.(The creation of the resource group does not rely on the watching etcd mechanism. Instead, it is automatically created after the first query for that resource group is dispatched to TiFlash.)

Author: ```suggestion
2. Watching GAC etcd, updating resource group configurations and detecting resource group deletions.(The creation of the resource group does not rely on the watching etcd mechanism. Instead, it is automatically created after the first query of that resource group is dispatched to TiFlash.)
```

2. Change the `RU_PER_SEC` of rg1 to 1000, observe CPU usage, and query execution times. Expect reduced CPU usage and increased query execution times for rg1.
3. Change the `RU_PER_SEC` of rg2 to 1000, observe CPU usage, and query execution times. Expect reduced CPU usage and increased query execution times for rg2.
4. Set the burstable attribute of rg1 to true, observe CPU usage, and query execution times. Expect increased CPU usage and significantly reduced query execution times for rg1.
5. Test that LAC reports accurate RU usage to GAC.
6. Test that the conversion between the three TokenBucket modes works as expected.
7. Test with multiple TiFlash nodes, running the above tests (1/2/3) with tpch workload, expecting similar results.
8. Test repeated deletion and reconstruction of resource groups. Expect queries to fail with "resource group not found" errors.

## Impacts & Risks
1. The introduction of ResourceControlQueue may lead to performance degradation as TaskScheduler's scheduling adds an extra layer of TaskQueue.

> 2. TiFlash resource control only works when TiDB system variable `tidb_enable_resource_control` and TiFlash config `enable_resource_control` are both true. Other, flow control and priority scheduling will not take effects.

Reviewer: ```suggestion
2. TiFlash resource control only works when TiDB system variable `tidb_enable_resource_control` and TiFlash config `enable_resource_control` are both true. Otherwise, flow control and priority scheduling will not take effects.
```
